Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs

```diff
@@ -374,12 +381,23 @@ endRide handle@ServiceHandle {..} rideId req = withLogTag ("rideId-" <> rideId.g
 
                 let (tollCharges, tollNames, tollConfidence) = do
                       let distanceCalculationFailure = distanceCalculationFailed || (maybe False (> 0) updRide.numberOfSelfTuned)
-                      if isJust updRide.estimatedTollCharges && isJust updRide.tollCharges && distanceCalculationFailure
-                        then (updRide.tollCharges, updRide.tollNames, Just Neutral)
-                        else
-                          if (isJust updRide.estimatedTollCharges || isJust updRide.tollCharges) && distanceCalculationFailure
-                            then (Nothing, Nothing, Just Unsure)
-                            else (updRide.tollCharges, updRide.tollNames, Just Sure)
+                      if distanceCalculationFailure
+                        then
+                          if isJust updRide.estimatedTollCharges
+                            then
+                              if isJust updRide.tollCharges
+                                then (updRide.tollCharges, updRide.tollNames, Just Neutral)
+                                else
+                                  if updRide.driverDeviatedToTollRoute == Just True
+                                    then (updRide.estimatedTollCharges, updRide.estimatedTollNames, Just Neutral)
+                                    else (Nothing, Nothing, Just Unsure)
+                            else (Nothing, Nothing, Just Unsure)
+                        else (updRide.tollCharges, updRide.tollNames, Just Sure)
+
+                fork "ride-interpolation" $ do
+                  interpolatedPoints <- getInterpolatedPoints updRide.driverId
+                  let rideInterpolationData = RideInterpolationData {interpolatedPoints = interpolatedPoints, rideId = updRide.id}
+                  when (isJust updRide.driverDeviatedToTollRoute && tollConfidence == Just Sure && maybe True (> 0) updRide.tollCharges && isJust updRide.estimatedTollCharges) $ pushToKafka rideInterpolationData "ride-interpolated-waypoints" updRide.id.getId
```

Comment:
- Replace `updRide.tollCharges` to `tollCharges` as it would always be Nothing at this point.